src/engine/GroupBy.cpp

```diff
@@ -956,41 +953,50 @@ void GroupBy::extractValues(
 }
 
 // _____________________________________________________________________________
+template <size_t GROUPED_COLUMNS>
 sparqlExpression::VectorWithMemoryLimit<ValueId>
 GroupBy::getHashMapAggregationResults(
-    IdTable* resultTable, const HashMapAggregationData& aggregationData,
+    IdTable* resultTable,
+    const HashMapAggregationData<GROUPED_COLUMNS>& aggregationData,
     size_t dataIndex, size_t beginIndex, size_t endIndex,
     LocalVocab* localVocab) {
   sparqlExpression::VectorWithMemoryLimit<ValueId> aggregateResults(
       getExecutionContext()->getAllocator());
   aggregateResults.resize(endIndex - beginIndex);
 
-  decltype(auto) groupValues = resultTable->getColumn(0);
   auto& aggregateDataVariant =
       aggregationData.getAggregationDataVariant(dataIndex);
 
-  auto op = [&aggregationData, &aggregateDataVariant, localVocab](Id val) {
-    auto index = aggregationData.getIndex(val);
-
-    auto visitor = [&index, localVocab](auto& aggregateDataVariant) {
-      return aggregateDataVariant.at(index).calculateResult(localVocab);
+  using B = HashMapAggregationData<GROUPED_COLUMNS>::template T<Id>;
```

Comment:
The cleanest way would be to also store a 
`VectorOrArray` of the columns (the spans of the IDTable)`, s.t. you can write
```
for (const auto& col : groupByColumnSpans) {
   target[currentIdx] = col[currentIdx] 
}
```.
That way this inner very tight for loop also benefits from the `std::array` optimization for very few columns.